### PR TITLE
Lock migration UI

### PR DESCRIPTION
**Pull request by clemsos**
# Description

This adds an UI to upgrade a lock from the user dashboard in `unlock-app`

**UI samples**

Screenshot 2021-12-09 at 16 18 04

Screenshot 2021-12-13 at 12 49 11

Screenshot 2021-12-13 at 12 48 15

****

# Issues

Fixes #
Refs #

# Checklist:

- [ ] 1 PR, 1 purpose: my Pull Request applies to a single purpose
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the docs to reflect my changes if applicable
- [ ] I have added tests (and stories for frontend components) that prove my fix is effective or that my feature works
- [ ] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

File: unlock-app/src/components/creator/lock/AppStore.js [hunk 1]
```diff
@@ -153,6 +153,24 @@ const AppStore = ({ lock }) => {
   return (
     <Wrapper>
       <Details>
+        {
+          lock.network === 100 || lock.network === 134 || lock.network === 4 // show on rinkeby for debug
```
Comment (julien51) [inline unlock-app/src/components/creator/lock/AppStore.js:157]: Ok, let's actually move this to `CreatorLock`. Also not all locks deployed on xdai or polygon are affected. only those for which the `unlock https://github.com/unlock-protocol/unlock/blob/c3f55c616f45e13dbcfd6a944ed5d3304d9cea2c/unlock-app/src/components/creator/CreatorLock.js#L117

File: newsletter/package.json [hunk 2]
```diff
@@ -25,7 +25,7 @@
     "eslint": "8.3.0",
     "jest": "27.4.3",
     "jest-date-mock": "1.0.8",
-    "next": "12.0.4",
+    "next": "^12.0.5-canary.8",
```
Comment (julien51) [inline newsletter/package.json:28]: Hum. Needed?

File: packages/unlock-js/src/v4/getLock.js [hunk 3]
```diff
@@ -79,5 +85,22 @@ export default async function (address, provider) {
   }
   delete update.tokenAddress
 
+  // check if compromised (Nov 22nd, 2021 attack)
+  if (update.unlockProtocol) {
```
Comment (julien51) [inline packages/unlock-js/src/v4/getLock.js:89]: I think we should do that check inside of unlock-app?

File: unlock-app/src/components/content/UpgradeContent.tsx [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+  return (
+    <Layout title="Upgrade Lock">
+      <Head>
+        <title>{pageTitle('Upgrade Lock')}</title>
+      </Head>
+      <Account />
+      <Heading>Upgrade your Lock</Heading>
+      <Instructions>
+        A new version of Unlock is available. Please click below to upgrade your
```
Comment (julien51) [inline unlock-app/src/components/content/UpgradeContent.tsx:55]:
```suggestion
        A new version of Unlock on {config.networks[network].name} is available. Please click below to clone your lock.
```

Comment (julien51) [reply unlock-app/src/components/content/UpgradeContent.tsx:55]: Please double check that this works :)

File: unlock-app/src/components/content/UpgradeContent.tsx [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+        lock and fix existing security issues.
```
Comment (julien51) [inline unlock-app/src/components/content/UpgradeContent.tsx:56]:
```suggestion
```

File: unlock-app/src/components/content/UpgradeContent.tsx [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+      </Instructions>
+      <p>
+        Note: An identical lock with a new address will be created, containing
```
Comment (julien51) [inline unlock-app/src/components/content/UpgradeContent.tsx:59]:
```suggestion
        Note: An identical lock with a new address will be created, with all members in an identical state.
```

File: unlock-app/src/components/content/UpgradeContent.tsx [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+        all existing content.{' '}
+        <b>
+          Once the migration has suceeded, please update your system with the
```
Comment (julien51) [inline unlock-app/src/components/content/UpgradeContent.tsx:62]:
```suggestion
          Once the migration has succeeded, please update your application with the new lock address. Feel free to reach out to the Unlock team on Discord for help.
```

File: unlock-app/src/components/content/UpgradeContent.tsx [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+          new lock address.
```
Comment (julien51) [inline unlock-app/src/components/content/UpgradeContent.tsx:63]:
```suggestion
```

File: unlock-app/src/components/content/UpgradeContent.tsx [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+        </b>
+      </p>
+      {error && <p>{error}</p>}
+      {!account && <p>Please authentificate to upgrade this lock</p>}
```
Comment (julien51) [inline unlock-app/src/components/content/UpgradeContent.tsx:67]:
```suggestion
      {!account && <p>Please authentificate to clone this lock</p>}
```

File: unlock-app/src/components/content/UpgradeContent.tsx [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+      {account && !error && (
+        <p>
+          <Button onClick={upgradeLock}>Upgrade your lock now</Button>
```
Comment (julien51) [inline unlock-app/src/components/content/UpgradeContent.tsx:70]:
```suggestion
          <Button onClick={upgradeLock}>Clone your lock now</Button>
```

File: unlock-app/src/components/creator/CreatorLock.js [hunk 11]
```diff
@@ -129,6 +128,19 @@ export const CreatorLock = ({
           .
         </LockWarning>
       )}
+      {lock.isCompromised && (
+        <LockWarning>
+          Your lock appears to be compromised, please{' '}
```
Comment (julien51) [inline unlock-app/src/components/creator/CreatorLock.js:133]:
```suggestion
          Your lock was deployed on an older Unlock contract, please{' '}
```

File: unlock-app/src/components/creator/CreatorLock.js [hunk 12]
```diff
@@ -129,6 +128,19 @@ export const CreatorLock = ({
...
+          <a
+            href={`/upgrade?locks=${lock.address}`}
+            target="_blank"
+            rel="noopener noreferrer"
+          >
+            upgrade to a recent version
```
Comment (julien51) [inline unlock-app/src/components/creator/CreatorLock.js:139]:
```suggestion
            clone it to use a new version of Unlock.
```

File: unlock-app/src/utils/exportStatic.js [hunk 13]
```diff
@@ -19,6 +19,7 @@ const exportPaths = (defaultPathMap, { dev, dir, outDir }) => {
     '/signup': { page: '/signup' },
     '/settings': { page: '/settings' },
     '/recover': { page: '/recover' },
+    '/upgrade': { page: '/upgrade' },
```
Comment (julien51) [inline unlock-app/src/utils/exportStatic.js:22]:
```suggestion
    '/clone': { page: '/clone' },
```

File: packages/unlock-js/src/v7/getLock.js [hunk 14]
```diff
@@ -79,5 +85,22 @@ export default async function (address, provider) {
   }
   delete update.tokenAddress
 
+  // check if compromised (Nov 22nd, 2021 attack)
+  if (update.unlockProtocol) {
+    const compromisedContracts = {
+      100: '0x3C6e461341AdbF7C0947085e86B4A6f35Ff2F801',
```
Comment (julien51) [inline packages/unlock-js/src/v7/getLock.js:91]: I think this is not the right address. The previous one was also `0x14bb3586ce2946e71b95fe00fc73dd30ed830863` on xdai,

File: unlock-app/src/__tests__/utils/exportStatic.test.js [hunk 15]
```diff
@@ -20,6 +20,7 @@ describe('exportStatic', () => {
       '/keychain': { page: '/keychain' },
       '/login': { page: '/login' },
       '/signup': { page: '/signup' },
+      '/upgrade': { page: '/upgrade' },
```
Comment (julien51) [inline unlock-app/src/__tests__/utils/exportStatic.test.js:23]:
```suggestion
      '/clone': { page: '/clone' },
```

File: unlock-app/src/components/content/UpgradeContent.tsx [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+    if (typeof fetch !== 'undefined' && network) {
+      try {
+        const response = await fetch(
+          `${config.networks[network].locksmith}/lock/${lockAddress}/migrate`,
+          { method: 'POST' }
+        )
+        console.log(await response.json())
+      } catch (error: any) {
+        console.log(error)
+        setError('Fail to upgrade. Please refresh and try again.')
```
Comment (julien51) [inline unlock-app/src/components/content/UpgradeContent.tsx:32]:
```suggestion
        setError('Fail to clone. Please refresh and try again.')
```

File: unlock-app/src/components/content/UpgradeContent.tsx [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+    }
+  })
+
+  return (
+    <Layout title="Upgrade Lock">
+      <Head>
+        <title>{pageTitle('Upgrade Lock')}</title>
+      </Head>
+      <Account />
+      <Heading>Upgrade your Lock</Heading>
```
Comment (julien51) [inline unlock-app/src/components/content/UpgradeContent.tsx:53]:
```suggestion
      <Heading>Clone your Lock</Heading>
```

**Review by julien51 — COMMENTED** [review]
Let's use `clone` rather than `upgrade` because I want to make sure we can use Upgrade later when we have actual upgrades!

**Comment by clemsos** [comment]
- [] pass signee at url level 
- [] show spinner if migration already happened (with contact us message)
- [] move migration to subprocess
- [] get call every 10s

File: packages/unlock-js/src/v4/getLock.js [hunk 18]
```diff
@@ -29,6 +30,11 @@ export default async function (address, provider) {
     publicLockVersion: parseInt,
   }
 
+  // fetch unlock contract address only on Polygon and Xdai (and rinkeby for testing)
+  if (chainId === 100 || chainId === 137 || chainId === 4) {
```
Comment (julien51) [inline packages/unlock-js/src/v4/getLock.js:34]:
```suggestion
  if (chainId === 100 || chainId === 137) {
```

File: packages/unlock-js/src/v6/getLock.js [hunk 19]
```diff
@@ -79,6 +85,23 @@ export default async function (address, provider) {
   }
   delete update.tokenAddress
 
+  // check if compromised (Nov 22nd, 2021 attack)
+  if (update.unlockProtocol) {
```
Comment (julien51) [inline packages/unlock-js/src/v6/getLock.js:89]: Let's move this logic to `unlock-app` please :)

File: unlock-app/package.json [hunk 20]
```diff
@@ -53,7 +53,7 @@
     "jest-mock-axios": "4.4.1",
     "jest-styled-components": "7.0.8",
     "js-base64": "3.7.2",
-    "next": "12.0.5",
+    "next": "^12.0.5-canary.8",
```
Comment (julien51) [inline unlock-app/package.json:56]: required?

File: unlock-app/src/pages/clone.js [hunk 21]
```diff
@@ -0,0 +1,11 @@
+import React from 'react'
+import { useRouter } from 'next/router'
+
+import CloneContent from '../components/content/CloneContent'
+
+const Upgrades = () => {
```
Comment (julien51) [inline unlock-app/src/pages/clone.js:6]:
```suggestion
const Clone = () => {
```

File: unlock-app/src/pages/clone.js [hunk 22]
```diff
@@ -0,0 +1,11 @@
...
+  const { query } = useRouter()
+  return <CloneContent query={query} />
+}
+
+export default Upgrades
```
Comment (julien51) [inline unlock-app/src/pages/clone.js:11]:
```suggestion
export default Clone
```

**Review by julien51 — COMMENTED** [review]
Let's move the logic of checking what Unlock contract is used to `unlock-app` please :)

File: unlock-app/package.json (hunk 20 above)
Comment (clemsos) [reply unlock-app/package.json]: ok removed

**Comment by clemsos** [comment]
> Let's move the logic of checking what Unlock contract is used to `unlock-app` please :)

yes this is using `previousVersions` field from the networks lib now

File: packages/unlock-js/src/v6/getLock.js (hunk 19 above)
Comment (clemsos) [reply packages/unlock-js/src/v6/getLock.js:84]: yep just kept the renaming

**Review by julien51 — APPROVED** [review]
💯 Let's do it!

**Pull request merged**